File: python/lombokalgoritma/graph/tarjan_scc.py
# LombokAlgoritma — strongly connected components, Tarjan (SPEC §9.10)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

from .types import Graph, build_adj_list


def tarjan_scc(g: Graph) -> list[list[int]]:
    """SCCs in completion order (reverse topological order of the condensation), members sorted.

    Iterative, but visits exactly like the recursive version (roots ascending, edge-list order).
    """
    adj = build_adj_list(g)
    n = len(adj)
    index = [-1] * n
    low = [0] * n
    on_stack = [False] * n
    stack: list[int] = []
    out: list[list[int]] = []
    counter = 0
    for root in range(n):
        if index[root] != -1:
            continue
        call = [[root, 0]]
        index[root] = low[root] = counter
        counter += 1
        stack.append(root)
        on_stack[root] = True
        while call:
            frame = call[-1]
            v, pos = frame
            edges = adj[v]
            if pos < len(edges):
                frame[1] = pos + 1
                w = edges[pos][0]
                if index[w] == -1:
                    index[w] = low[w] = counter
                    counter += 1
                    stack.append(w)
                    on_stack[w] = True
                    call.append([w, 0])
                elif on_stack[w]:
                    low[v] = min(low[v], index[w])
                continue
            call.pop()
            if call:
                parent = call[-1][0]
                low[parent] = min(low[parent], low[v])
            if low[v] == index[v]:
                comp: list[int] = []
                while True:
                    w = stack.pop()
                    on_stack[w] = False
                    comp.append(w)
                    if w == v:
                        break
                out.append(sorted(comp))
    return out

File: rust/lombokalgoritma/src/graph/tarjan_scc.rs
// LombokAlgoritma — Tarjan strongly connected components (SPEC §9.10)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
use super::types::{adjacency, Graph};
use crate::Result;
use alloc::vec;
use alloc::vec::Vec;

/// Strongly connected components (Tarjan), iterative but visiting exactly like the recursive
/// version: roots ascending, neighbours in edge-list order. Components are returned in completion
/// order (reverse topological order of the condensation), each sorted ascending. O(V + E).
///
/// # Errors
/// [`crate::Error::OutOfRange`] for an edge endpoint outside `[0, n)`.
pub fn tarjan_scc(g: &Graph) -> Result<Vec<Vec<usize>>> {
    let adj = adjacency(g)?;
    let n = g.nodes;
    let mut index: Vec<Option<usize>> = vec![None; n];
    let mut low = vec![0usize; n];
    let mut on_stack = vec![false; n];
    let mut stack: Vec<usize> = Vec::new();
    let mut out = Vec::new();
    let mut counter = 0usize;
    for root in 0..n {
        if index[root].is_some() {
            continue;
        }
        let mut call: Vec<(usize, usize)> = vec![(root, 0)];
        index[root] = Some(counter);
        low[root] = counter;
        counter += 1;
        stack.push(root);
        on_stack[root] = true;
        while let Some(&(v, pos)) = call.last() {
            if pos < adj[v].len() {
                if let Some(frame) = call.last_mut() {
                    frame.1 = pos + 1;
                }
                let w = adj[v][pos].0;
                match index[w] {
                    None => {
                        index[w] = Some(counter);
                        low[w] = counter;
                        counter += 1;
                        stack.push(w);
                        on_stack[w] = true;
                        call.push((w, 0));
                    }
                    Some(iw) if on_stack[w] => low[v] = low[v].min(iw),
                    Some(_) => {}
                }
                continue;
            }
            call.pop();
            if let Some(&(parent, _)) = call.last() {
                low[parent] = low[parent].min(low[v]);
            }
            if Some(low[v]) == index[v] {
                let mut comp = Vec::new();
                while let Some(w) = stack.pop() {
                    on_stack[w] = false;
                    comp.push(w);
                    if w == v {
                        break;
                    }
                }
                comp.sort_unstable();
                out.push(comp);
            }
        }
    }
    Ok(out)
}
